sql/catalog/hydratedtables: add a cache for hydrated table descriptors and adopt

ajwerner opened the pull request:
See individual commits.

Fixes #51385.

Release note: None

Commit:
sql/catalog/hydratedtables: add a cache for hydrated table descriptors

This cache ends up being important for dealing with user-defined types.

Release note: None

Release justification:

Commit:
sql: adopt hydratedtables

Release note: None

Commit:
schemadesc: rename NewMutableCreatedSchemaDescriptor to NewCreateedMutable

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53323)
<!-- Reviewable:end -->

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache_test.go:
```diff
@@ -0,0 +1,341 @@
+	typ2Name        = tree.MakeNewQualifiedTypeName(dbDesc.Name, schemaDesc.Name, typ2Desc.Name)
+	typ2T           = mkTypeT(typ2Desc, &typ2Name)
+	typ2TSerialized = &types.T{InternalType: typ2T.InternalType}
+	tableDescUDF    = tabledesc.NewCreatedMutable(descpb.TableDescriptor{
```
nit, but what does UDF stand for here

ajwerner replied:
whoops mean UDT, user-defined type.

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache_test.go:
```diff
@@ -0,0 +1,341 @@
+		// type one time.
+		require.Equal(t, res.calls, 2)
+	})
+	t.Run("no udf, no metrics", func(t *testing.T) {
```
empty test

ajwerner replied:
Done.

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
Can you explain what you ran into that needs this behavior?

ajwerner replied:
Yeah, commented

ajwerner commented on pkg/sql/catalog/descs/collection.go:
```diff
@@ -154,14 +156,22 @@ func MakeCollection(
+	hydratedTables *hydratedtables.Cache,
 ) *Collection {
-	tc := MakeCollection(ctx, leaseMgr, settings, nil, nil, nil)
+	tc := MakeCollection(ctx, leaseMgr, settings,
```
Done.

ajwerner commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
```diff
@@ -0,0 +1,265 @@
+// caller needs to have a lease on the relevant type descriptor. The way that
+// this is made to work is that the user provides a handle to a leased
+// ImmutableCopy and then the cache will call a function for each
+// of the
```
Done.

ajwerner commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
```diff
@@ -0,0 +1,265 @@
+	io_prometheus_client "github.com/prometheus/client_model/go"
+)
+
+// TODO(ajwerner): Add metrics.
```
Removed.

ajwerner commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
```diff
@@ -0,0 +1,265 @@
+		settings: settings,
+		metrics:  makeMetrics(),
+	}
+	c.mu.cache = cache.NewUnorderedCache(cache.Config{
```
This one isn't very good though :/

ajwerner commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
Done.

ajwerner commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
Done but I'm not totally clear on why it matters. 

> we want to store the types in a slice in the order they are resolved in the cachedTypeDescriptorResolver

Why?

rohany replied:
Never mind, I was still in the mindset that the types were going to be part of the key. Either way, its good to be deterministic.

ajwerner commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
indeed, I somewhat wish their key was `interface{}` and not `string` though

ajwerner commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
I don't disagree but we can tie that together later. Making the caching work for the batched retrieval is a whole other thing.

rohany replied:
Yeah, just pointing it out

ajwerner commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
Done. Pointer equality worked in my head because we get the same pointers out of the lease manager but I think you're spot on that this will spell trouble for the distsql flows.

ajwerner commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
I actually had typed it that way first but I wanted to actively avoid allocating objects on the fast path. I went and checked and discovered key was escaping so I fixed that.

ajwerner commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
That's what I mean by finer granularity. It will avoid the thrashing but it's a good deal more complex. In particular, it's more complex w.r.t. the cache. I'll do it as a follow-up if you'll allow it.

ajwerner commented on pkg/sql/catalog/hydratedtables/hydratedcache_test.go:
```diff
@@ -0,0 +1,341 @@
+)
+
+func TestHydratedCache(t *testing.T) {
+	defer leaktest.AfterTest(t)()
```
Huh?

rohany replied:
I thought that all of the tests needed this too 

```
	defer log.Scope(t).Close(t)
```

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
this is unused

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
I'm still confused. Right now, the cache isn't accessed by DistSQL (we only resolve types through the lease manager, unless you're planning for the future now). Based on the code above, if this returns false, we'll still go and put the descriptor into the cache with the same key, so we would thrash right?

ajwerner replied:
right, right there's a bit more to do here. I'm not sure I follow what you mean about thrashing. It shouldn't thrash back and forth. Sure a resolved name should kick out an unresolved name but not vice-versa. 

I've had two different approach I've been thinking about:

1) have distsql resolve the table using the machinery (except in the case where it's a modified table) but in that case I believe the distsql stuff is set up to not
2) Augment the cache built here to have a cache for the underlying `*types.T` values and utilize those when hydrating descriptors to avoid just a few allocs. 

1) is better but bigger is my sense. In either case, we want the above logic I think.

ajwerner commented:
Good new is I benchmarked this and found:

```
name             old ops/s   new ops/s   delta
KV95-throughput  14.4k ± 2%  16.5k ± 1%  +14.52%  (p=0.008 n=5+5)

name             old ms/s    new ms/s    delta
KV95-P50          13.3 ± 2%   11.5 ± 0%  -13.53%  (p=0.000 n=5+4)
KV95-Avg          3.44 ± 2%   3.10 ± 0%   -9.88%  (p=0.008 n=5+5)
```

Still working though the logictest changes

rohany commented:
nice, glad to see this measurable impact

Commit:
catalog: extend Descriptor to include IsUncommittedVersion

This is handy when trying to determine whether we can cache a descriptor.

Release justification: Part of a change to eliminate a major performance
degradation for a new feature.

Release note: None

Commit:
sql/types,catalog/typedesc: don't re-hydrate types

Adds an IsHydrated method to types and then uses it to determine whether a type
should be hydrated.

Release note: None

Commit:
tabledesc,descs: ensure to propagate IsUncommittedVersion when hydrating

When we hydrate a table descriptor which has been modified during the
transaction we want to ensure that we propagate the information that
indicates that it has been modified.

Release justification: Part of a larger fix to a major performance
enhancement for user-defined types, a new feature in this release.

Release note: None

ajwerner commented:
Okay @rohany I did less drastic things than at one point we discussed. I think this now works PTAL.

rohany commented on pkg/sql/catalog/dbdesc/database_desc.go:
```diff
@@ -31,6 +31,10 @@ var _ catalog.MutableDescriptor = (*Mutable)(nil)
 type Immutable struct {
 	descpb.DatabaseDescriptor
+
+	// isModified is set to true if this descriptor was created a copy of a
```
*created from a copy of

ajwerner replied:
Done.

rohany commented on pkg/sql/catalog/dbdesc/database_desc.go:
```diff
@@ -110,6 +114,11 @@ func (desc *Immutable) GetParentID() descpb.ID {
 }
 
+// IsModified implements the Descriptor interface.
+func (desc *Immutable) IsModified() bool {
```
this is a really weird interface, (maybe I'll think of an alternative by the end of the review). It doesn't make any sense for an Immutable descriptor to ever return true to a "IsModified" API

ajwerner replied:
I disagree

Imagine 
```sql
CREATE TABLE t(i INT);
BEGIN;
ALTER TABLE t DROP COLUMN I;
SELECT * FROM t; -- this should resolve a modified immutable descriptor 
```

rohany replied:
I understand that it could happen, but I don't think that the verbiage makes sense.

ajwerner replied:
Okay so your gripe isn't the implementation but rather the name and the commenting, yeah? If I went with `IsUncommittedVersion()` would that make it clearer? Also add some commentary about how it works and about the settings in which it is reliable and put those in the interface comment?

rohany replied:
Yeah. I can't think of a different way to do what you're trying to do.

rohany commented on pkg/sql/catalog/descriptor.go:
```diff
@@ -43,6 +43,10 @@ type Descriptor interface {
 
+	// IsModified returns true if this descriptor does not represent a currently
+	// committed version.
+	IsModified() bool
```
Maybe `IsUncommitted`?

ajwerner replied:
Done

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
unfinished comment

ajwerner replied:
Done.

rohany reviewed (commented):
Going to take another pass at this tonight, but left some preliminary comments

rohany commented on pkg/sql/catalog/descs/collection.go:
```diff
@@ -1078,14 +1093,25 @@ func (tc *Collection) hydrateTypesInTableDesc(
 			return name, desc, nil
+		})
+
+		// Utilize the cache of hydrated tables if we have one.
```
Add a comment about the fact that we might get rejected from the cache if we have an uncommitted descriptor, or if any of the types referenced are uncommitted too

ajwerner replied:
Done.

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
comment this

ajwerner replied:
Done.

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
for clarity, I'd spell out why you need the for

ajwerner replied:
Done.

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
I think that you only use `cached` in this scope, so maybe make it explicit `if cached, ok := ...`

ajwerner replied:
Done

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
I'm not sure that this is going to work for you -- there are cases where the `err` variable is getting shadowed and won't set this. Should wrap the main logic in a function or lambda here and deal with the response

ajwerner replied:
It doesn’t matter if it’s in scope or shadowed or whatever. When you have a named return value it gets set to the return value in the return statement.

rohany replied:
Huh, TIL that is some strange behavior...

ajwerner replied:
I think of this as a pretty critical feature in the who `defer` scheme. It's very important to be able to rely on the idea that when you're running a defer you'll be able to see the values being returned from the function (unless they are modified by an earlier return). 

One handy trick is

```
func f(args interface{}) (ret interface{}, err error) {
    defer func() {  log.Infof(..., args, ret, err) }(args)
}
```

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
comment this

ajwerner replied:
Done.

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
why does this matter

ajwerner replied:
because it's in the data structure, commented

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache.go:
im more sold on uncommitted vs modified terminology now

rohany commented on pkg/sql/catalog/hydratedtables/hydratedcache_test.go:
```diff
@@ -0,0 +1,421 @@
+		// This case should not meaningfully arise in practice.
+		td := tableDescUDT.ImmutableCopy().(*tabledesc.Immutable)
+		{
+			hydrated, canUse, err := c.GetHydratedTableDescriptor(ctx, td, resWithMut)
```
why can we use cache in this case and below? Isn't the type modified

ajwerner replied:
Can use does not mean that we definitely stored it in the cache, just that we hydrated it underneath the call.

ajwerner commented:
TFTR!

bors r=rohany
